### PR TITLE
add rvalues matching to ref parameters

### PR DESCRIPTION
**Pull request by WalterBright**
Although the DIP isn't finished yet, we can try out the ideas.

### CODE REVIEW

**Comment by jacob-carlborg** [comment]
Next time, please do a draft pull request: https://github.blog/2019-02-14-introducing-draft-pull-requests.

**Comment by WalterBright** [comment]
Currently blocked by https://github.com/dlang/phobos/pull/7015

**Comment by thewilsonator** [comment]
Merged.

File: test/runnable/structlit.d
Comment (thewilsonator) [inline test/runnable/structlit.d]: Move this down two lines

File: test/runnable/template9.d
Comment (thewilsonator) [inline test/runnable/template9.d]: Presumably this no longer causes an error

File: test/runnable/testassign.d
Comment (thewilsonator) [inline test/runnable/testassign.d]: Does this still apply?

**Comment by thewilsonator** [comment]
In general all those `static assert(__traits(compiles, foo());` should just be `foo();`

**Comment by thewilsonator** [comment]
One more phobos case https://github.com/dlang/phobos/blob/master/std/parallelism.d#L3873-L3887

File: test/fail_compilation/ice8255.d
Comment (Geod24) [inline test/fail_compilation/ice8255.d]: That error message should be improved

**Review by wilzbach — DISMISSED** [review]
Should be behind a -preview flag, s.t. it can be merged and tested quickly.

**Comment by aliak00** [comment]
So this breaks a `NotNull` type. The purpose of which is to have a guarantee that your type is not null. Without this patch, this would fail:
```
struct NotNull(T) {
  private T _value;
  @property ref inout(T) value() inout { return this._value; }
  alias value this;
}

void func(ref T t) {
  t = null;
}

auto a = notNull!T;
func(a); // compile error
```

This patch makes it pass. So I could remove the alias this. But then NotNull becomes useless because you can't use it as a `T`.

Is there a way I can fix this or work around it?

@thewilsonator pointed to the failing unittest: https://github.com/aliak00/optional/blob/master/tests/notnull.d#L105

**Comment by WalterBright** [comment]
> Is there a way I can fix this or work around it?

Disable copying NotNull by declaring a postblit and annotate it with @disable.

**Comment by aliak00** [comment]
> Disable copying NotNull by declaring a postblit and annotate it with @disable.

That's way too crippling. You can't pass it around then. Another option is removing the `ref` from the property, but then you can't manipulate any fields, so it also cripples it too much :(

**Comment by atilaneves** [comment]
@aliak00 I'm not sure what you mean. This compiles today:

```d
struct NotNull(T) {
  private T _value;
  @property ref inout(T) value() inout { return this._value; }
  alias value this;
}

void func(ref Class t) {
    t = null;
}

class Class {}

void main() {
    auto a = NotNull!Class();
    func(a); 
}
```

Did I misunderstand your code?

**Comment by aliak00** [comment]
> Did I misunderstand your code?

Nope. And yeah you're right :/ It was only the rvalue case that didn't compile, i erroneously assumed the same was with the lvalue case. I guess a `NotNull` is going to be harder than I thought 🎉

**Comment by thewilsonator** [comment]
In the interest of moving forward with this I shall rescind my minor quibbles about misleading comments in the test suite

**Comment by thewilsonator** [comment]
This should probably have a changelog as some point, too.

**Comment by Geod24** [comment]
> Although the DIP isn't finished yet

I would have imagined that, even for `-preview`, the design should be done before a merge ?

**Comment by thewilsonator** [comment]
Well, for fixing shared I want to do something similar, namely disable `shared` read and write access under a preview flag in @safe cod, that while the full design of the DIP is not finished (what with memory models and so on), is necessary in every conceivable solution to `shared`. 

That way Manu can do the exploration to see what breaks and needs fixing. That applies here too ;)

**Comment by wilzbach** [comment]
> I would have imagined that, even for `-preview`, the design should be done before a merge ?

Perfect is the enemy of good and progress. The preview flags are intended for exactly this reason - designs which aren't finished yet and can be experimented with.

**Comment by thewilsonator** [comment]
@WalterBright please rebase.

**Comment by Geod24** [comment]
Needs a rebase. Also, what's the plan for `auto ref` after this ?

**Comment by TurkeyMan** [comment]
Andrei described 2 approaches to the ambiguous case resolution in his dconf lecture. Which of the 2 approaches is implemented?
I think only the second approach was valid.

**Comment by WalterBright** [comment]
This one was the different conversion level.

**Comment by TurkeyMan** [comment]
... I think that's the second one? I only remember them on chronological order.

**Comment by Geod24** [comment]
All those tests change should *really* have gone to their own files instead of polluting the existing ones. Changing existing tests should really not be done when adding a `-preview` switch, because it is essentially removing tests without the switch.

For reference, this implementation is incomplete ([it doesn't work for the most trivial template](https://issues.dlang.org/show_bug.cgi?id=20705) among other things), and while trying to fix it I realized I needed to alter a bunch of tests, but that means the old (sane) behavior won't be tested anymore.

**Comment by thewilsonator** [comment]
template issues fixed in https://github.com/dlang/dmd/pull/12318

**Pull request merged**
